runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/environment/DockerEnvironmentFactory.java

```diff
@@ -0,0 +1,153 @@
+      GrpcFnServer<StaticGrpcProvisionService> provisioningServiceServer,
+      ControlClientSource clientSource,
+      // TODO: Refine this to IdGenerator when we determine where that should live.
+      Supplier<String> idGenerator) {
+    return new DockerEnvironmentFactory(
+        docker,
+        controlServiceServer,
+        loggingServiceServer,
+        retrievalServiceServer,
+        provisioningServiceServer,
+        idGenerator,
+        clientSource);
+  }
+
+  private final DockerWrapper docker;
+  private final GrpcFnServer<FnApiControlClientPoolService> controlServiceServer;
+  private final GrpcFnServer<GrpcLoggingService> loggingServiceServer;
+  private final GrpcFnServer<ArtifactRetrievalService> retrievalServiceServer;
+  private final GrpcFnServer<StaticGrpcProvisionService> provisioningServiceServer;
+  private final Supplier<String> idGenerator;
+  private final ControlClientSource clientSource;
+
+  private RemoteEnvironment dockerEnvironment = null;
+
+  private DockerEnvironmentFactory(
+      DockerWrapper docker,
+      GrpcFnServer<FnApiControlClientPoolService> controlServiceServer,
+      GrpcFnServer<GrpcLoggingService> loggingServiceServer,
+      GrpcFnServer<ArtifactRetrievalService> retrievalServiceServer,
+      GrpcFnServer<StaticGrpcProvisionService> provisioningServiceServer,
+      Supplier<String> idGenerator,
+      ControlClientSource clientSource) {
+    this.docker = docker;
+    this.controlServiceServer = controlServiceServer;
+    this.loggingServiceServer = loggingServiceServer;
+    this.retrievalServiceServer = retrievalServiceServer;
+    this.provisioningServiceServer = provisioningServiceServer;
+    this.idGenerator = idGenerator;
+    this.clientSource = clientSource;
+  }
+
+  /** Creates an active {@link RemoteEnvironment} backed by a local Docker container. */
+  @Override
+  public RemoteEnvironment createEnvironment(Environment environment) throws Exception {
+    if (dockerEnvironment == null) {
+      dockerEnvironment = createDockerEnv(environment);
+    } else {
+      checkArgument(
+          environment.getUrl().equals(dockerEnvironment.getEnvironment().getUrl()),
+          "A %s must only be queried for a single %s. Existing %s, Argument %s",
+          DockerEnvironmentFactory.class.getSimpleName(),
+          Environment.class.getSimpleName(),
+          dockerEnvironment.getEnvironment().getUrl(),
+          environment.getUrl());
+    }
+    return dockerEnvironment;
+  }
+
+  private DockerContainerEnvironment createDockerEnv(Environment environment) throws Exception {
+    String workerId = idGenerator.get();
+
+    // Prepare docker invocation.
+    Path workerPersistentDirectory = Files.createTempDirectory("worker_persistent_directory");
+    Path semiPersistentDirectory = Files.createTempDirectory("semi_persistent_dir");
+    String containerImage = environment.getUrl();
+    // TODO: The default service address will not work for Docker for Mac.
```

Comment:
<!--thread_id:cc_182896308_t; commit:eb0cdf4ea2d094d8ab09a0a9ebd742507f71a1a8; resolved:1-->
<!--section:context-quote-->
> **jkff** wrote:
> Does that mean this whole thing (Flink portable runner) doesn't currently work on Mac?

<!--section:body-->
Not quite. We ended up switching the server factory in the hacking branch to switch endpoint urls based on OS. I'll add a bug for that.